### ISSUE TITLE
Update error representation for Failables

### ISSUE THREAD

**Issue by lerno** [c3lang/c3c#259]
## Background

Today error values are declared this way:

    errtype MyError; // Short error
    
    errtype MyParseError 
    {
      uint line;
    }
    
Using these:

    if (...) return MyError!;
    if (...) return MyParseError({ current_line });

Errors may be stored as such:

    MyParseError x = MyParseError({ 123 });

Or placed in a union which can contain any error:

    anyerr z = MyError;
    if (x > 0) z = MyParseError({ 123 });
    
`anyerr` is a struct `{ iptr, iptr }`. The first part is a typeid containing the type, whereas the second part contains
the data inside of the error. This limits the size of an error struct to be iptr wide.
            
So in short:

1. Errors are structs that have a size limited (or padded) to iptr width.
2. `anyerror` can contain any error, by virtue of containing the typeid + the data.
3. To inspect an error inside, one must first use the typeid to find the error type, then the data may be accessed.

A common way to create error domains today would be to use this strategy:

    enum IoErrorKind
    {
        NO_SUCH_FILE,
        INVALID_PATH,
        ...
    }
        
    errtype IoError
    {
       IoErrorKind error_kind;
       int optional_data;
    }
    
    ... return IoError({ .error_kind = NO_SUCH_FILE })!;

**Comment by lerno** [comment]
## Proposal

Instead of using `{ iptr, iptr }` the return is simply `iptr` sized. The following check is performed on this value:

1. If `(iptr & 0x01) == 0x01` then this is, when the lowest bit is removed, a pointer to an error `ErrorValue*`
2. Otherwise iptr is the error code.

The ErrorValue starts with a `typeid` allowing identification of the error domain.

We can now distinguish two error types:

1. Error types that are enums given a domain.
2. Error types that are unique structs.

The question is how the syntax should look like, esp with added data.

Let's say we have some struct which we want to pass as a pointer:

```c
struct ParserError : errtype <- this includes a typeid
{
   char* message;
   int line;
   int column;
   File* file;
}

void! parse()
{
   ...
   if (!x) {
     ParserError *error = mem::malloc(ParserError);
     *error = { .messages = "X is missing", .line = line, .column = col, .file = file };
     return error!;
   }
}
```
Where is the typeid set? In this case possibly at `error!`, but in some cases the type will not be well known:

```c
ParserError *error = ...
anyerr e = error; // <- when assigning to anyerr, the typeid must be set as well.
```
    
For enums errors, this sort of syntax is a possibility:

```c
enum IoError : errtype
{
   FILE_NOT_FOUND,
   INVALID_PATH,
}
```

In this case the enum will retain all the normal enum functionality, except for not allowing values to be set, that is the values are ordinals.

Thus we'd get something like:

```c
if (path_is_invalid) return IoError.INVALID_PATH!;
if (next_char() != '?')
{
   ParserError *error = mem::malloc(ParserError);
   *error = { .messages = "X is missing", .line = line, .column = col, .file = file };
   return error!;
}
```

The latter can obviously be simplified with a function, so code would probably look more like:

```c
if (path_is_invalid) return IoError.INVALID_PATH!;
if (next_char() != '?') return parse::new_error("X is missing", line, col, file)!;
```

Other syntax variants are possible and it would be nice to get some proposals.

**Comment by lerno** [comment]
### Usages

The current way to catch errors by type looks
```c
func void! findFileAndParse2()
{
    catch (err = findFileAndParse())
    {
        case FileNotFoundError:
            printf("Error loading the file!\n");
        default:
            return err;
    }
}
```

We might want this to look differently in this proposal:

```c
func void! findFileAndParse2()
{
    catch (err = findFileAndParse())
    {
        case IoError.INVALID_PATH:
            ...
        case IoError.FILE_NOT_FOUND:    
           ...
        case ParseError:
           // Here we can use the parse error:
           printf("Parse error at line %d\n", err.line);              
        default:
            return err!;
    }
}
```

**Comment by lerno** [comment]
### Implementation concerns

All of the above can be used with a single `iptr`

*Example 1*
```c
if (!x) {
  ParserError *error = mem::malloc(ParserError);
  *error = { .messages = "X is missing", .line = line, .column = col, .file = file };
  return error!;
}
if (path_is_invalid) return IoError.INVALID_PATH!;

```

This lowers to:
```c
short[2] IoErrorBase @align(4);
const iptr IoErrorFileNotFound = (iptr)(&IoErrorBase[0]);
const iptr IoErrorInvalidPath = (iptr)(&IoErrorBase[1]);

...
if (!x) {
  ParserError *error = mem::malloc(ParserError);
  *error = { .messages = "X is missing", .line = line, .column = col, .file = file };  
  return {| error.typeid = ParserError.typeid; return (iptr)(error) | 0x01; |};
}
if (path_is_invalid) return IoErrorInvalidPath;
```

*Example 2*
```c
func void! findFileAndParse2()
{
    catch (err = findFileAndParse())
    {
        case IoError.INVALID_PATH:
            ...
        case IoError.FILE_NOT_FOUND:    
           ...
        case ParseError:
           // Here we can use the parse error:
           printf("Parse error at line %d\n", err.line);              
        default:
            return err!;
    }
}
```
It is lowered to this:
```c
func iptr findFileAndParse2()
{
    iptr err = findFileAndParse();
    if (err)
    {
       if (err == IoErrorInvalidPath) 
       {
          ...
       }   
       else if (err == IoErrorFileNotFound)
       {
          ...
       }
       else if ((err & 0x01) && ((typeid*)(err ^ 0x01)) == ParseError.typeid) {
          ParseError *err = (ParseError *)(err); // Shadowing
           printf("Parse error at line %d\n", err.line);
       } 
       else 
       {
          return err;
       }
    }                 
}
```

**Comment by lerno** [comment]
If this representation is updated, then most of the cases where you would use something like `while (next(&value)) { ... }` could be used without extra cost.

An example:

```c
func bool Iterator.next(Iterator *it, int  *value)
{
   if (it.index >= it.len) return false;
   *value = it.array[it.index++];
   return false;
}

func void test()
{
   ...
   int value;
   while (it.next(&value)) { ... };
```

With failable:
```c
func bool Iterator.next(Iterator *it, int  *value)
{
   if (it.index >= it.len) return false;
   *value = it.array[it.index++];
   return false;
}

func void test()
{
   ...
   int value;
   while (value = it.next() else break) { ... };
}   
```

If we look at the what the failable would lower to it would be:

```c
while (int temp; err = it.next(&temp), if (err) break, value = temp) { ... }
```

**Comment by lerno** [comment]
As an alternative the `0x01` can be completely removed by using this:

char[4] IoErrorBase @align(4);

const iptr IoErrorFileNotFound = (iptr)(&IoErrorBase[1]);
const iptr IoErrorInvalidPath = (iptr)(&IoErrorBase[3]);

This will ensure that the offset is always distinguishable from a pointer without a need to do add `| 0x01` anywhere. Plus they can be easily distinguishable by using `& 0x01`

**Comment by lerno** [comment]
To avoid the ad hoc enum/struct reuse:
```c
errtype IoError
{
  INVALID_PATH,
  FILE_NOT_FOUND,
  ILLEGAL_CHARACTER(char* name)
}

...
return IoError.INVALID_PATH!;
return IoError.ILLEGAL_CHARACTER("hello");

anyerr e = ...
if (e == IoError.ILLEGAL_CHARACTER) puts(e.name);
switch (e)
{
   case IoError.INVALID_PATH:
     ...
   case IoError.ILLEGAL_CHARACTER:
     puts(e.name);
}
```

**Comment by lerno** [comment]
Comparing:
Iterate using while with an "optional" response
```swift
// Swift
while let i = it.next() { /* ... */ }
```
```c
// C3
while (try int x = it.next()) { /* ... */ }
```
Call a function that may return an error and handle it, then return:
```swift
// Swift
guard let x = someError() else {
   /* ... */
   return;
}
/* .. use x ... */
```
```go
// Go
x, err := someError();
if (err != nil) { 
   /* ... */
   return;
}
```
```c
// C3
int! x = someError();
if (catch err = x) {
   /* ... */
   return;
}
/* ... use x here ... */
```
Do something only if a call succeeds using the result
```swift
// Swift
if let x = someError() {
   /* ... */
}
```
```go
// Go
x, err := someError();
if (err == nil) { 
  /* ... */
}
```
```c
// C3
if (try int x = someError()) {
  /* ... */
}   
```
Call two functions, chained that may error then handle the resulting error.
```swift
// Swift
guard let x = someError(), let y = someOtherCall(x) {
  /* handle fail */
}
```
```c
// C3
if (catch err = someOtherCall(someError()) {
  /* ... */
}   
```
Give two equally valid results:
```swift
// Swift
if let roomCount = john.residence?.numberOfRooms {
    print("John's residence has \(roomCount) room(s).")
} else {
    print("Unable to retrieve the number of rooms.")
}
```
```c
// C3
if (try int room_count = john.getResidence().numberOfRooms) {
    io::printf("John's residence has %d room(s).", room_count);
} else {
    io::println("Unable to retrieve the number of rooms.")
}
// Other possibility:
bool failed = catch io::printf("John's residence has %d room(s).",  john.getResidence().numberOfRooms);
if (failed) {
    io::println("Unable to retrieve the number of rooms.")
}
```

**Comment by lerno** [comment]
try / catch expressions:

`try (<failable-expr>)`
`catch (<failable-expr>)`

cond-expr changes into this:

```
cond-expr ::= ((expr | decl) COMMA)* cond-expr-end
cond-expr-end ::= expr | decl | try-catch-decl
try-catch-decl ::=  try-decl | catch-decl
try-decl ::= try-decl-part ((AND | OR) try-decl-part | not-and-expr)*
try-decl-part ::= TRY IDENT | TRY (type)? IDENT = not-and-expr
catch-decl ::= catch-decl-part ((AND | OR) catch-decl-part | not-and-expr)*
catch-decl-part ::= catch IDENT | catch (type)? IDENT = not-and-expr
```

**Comment by lerno** [comment]
This is implemented.
